x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/SourceOnlySnapshotEngine.java

```diff
@@ -0,0 +1,478 @@
+    private final SearcherManager searcherManager;
+    private final IndexCommit indexCommit;
+    private final Lock indexWriterLock;
+
+    public SourceOnlySnapshotEngine(EngineConfig config) {
+        super(config);
+        try {
+            Store store = config.getStore();
+            store.incRef();
+            DirectoryReader reader = null;
+            Directory directory = store.directory();
+            Lock indexWriterLock = null;
+            boolean success = false;
+            try {
+                indexWriterLock = directory.obtainLock(IndexWriter.WRITE_LOCK_NAME);
+                this.lastCommittedSegmentInfos = Lucene.readSegmentInfos(directory);
+                this.translogStats = new TranslogStats(0, 0, 0, 0, 0);
+                final SequenceNumbers.CommitInfo seqNoStats =
+                    SequenceNumbers.loadSeqNoInfoFromLuceneCommit(lastCommittedSegmentInfos.userData.entrySet());
+                long maxSeqNo = seqNoStats.maxSeqNo;
+                long localCheckpoint = seqNoStats.localCheckpoint;
+                this.seqNoStats = new SeqNoStats(maxSeqNo, localCheckpoint, localCheckpoint);
+                reader = SeqIdGeneratingDirectoryReader.wrap(ElasticsearchDirectoryReader.wrap(DirectoryReader
+                .open(directory), config.getShardId()), config.getPrimaryTermSupplier().getAsLong());
+                this.indexCommit = reader.getIndexCommit();
+                this.searcherManager = new SearcherManager(reader, new SearcherFactory());
+                this.indexWriterLock = indexWriterLock;
+                success = true;
+            } finally {
+                if (success == false) {
+                    IOUtils.close(reader, indexWriterLock, store::decRef);
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e); // this is stupid
+        }
+    }
+
+    @Override
+    protected void closeNoLock(String reason, CountDownLatch closedLatch) {
+        if (isClosed.compareAndSet(false, true)) {
+            try {
+                IOUtils.close(searcherManager, indexWriterLock, store::decRef);
+            } catch (Exception ex) {
+                logger.warn("failed to close searcher", ex);
+            } finally {
+                closedLatch.countDown();
+            }
+        }
+    }
+
+    @Override
+    public GetResult get(Get get, BiFunction<String, SearcherScope, Searcher> searcherFactory) throws EngineException {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    protected ReferenceManager<IndexSearcher> getReferenceManager(SearcherScope scope) {
+        return searcherManager;
+    }
+
+    @Override
+    protected SegmentInfos getLastCommittedSegmentInfos() {
+        return lastCommittedSegmentInfos;
+    }
+
+    @Override
+    public String getHistoryUUID() {
+        return lastCommittedSegmentInfos.userData.get(Engine.HISTORY_UUID_KEY);
```

Comment:
I am not sure why this matters?

Reply:
it matters because the semantics of a history uuid is that if two shards have the same uuid they can be assured that all ops below the global checkpoint are identical. This engine violates this so it will make sure that it's history uuid isn't equal to anything else. I don't see exactly how this will end up being a problem but I rather maintain semantics than worry about it.